File: packages/protocol-ts/src/serialization.test.ts
/**
 * Property 6 — Roundtrip serialization.
 *
 * For every protocol message, `parse(serialize(msg)) === msg`, including null
 * and absent-optional fields. Each property runs for at least 100 generated
 * inputs (Requirement 13.10).
 *
 * Validates: Requirements 8.1, 7.3, 7.5
 */

import fc from 'fast-check';
import { describe, expect, it } from 'vitest';
import {
  parseCapabilities,
  parseChunkRef,
  parseDeviceInfo,
  parseFileMeta,
  parsePrepareUploadRequest,
  parsePrepareUploadResult,
  parseTransferPlan,
  parseTransferProgress,
  serializeCapabilities,
  serializeChunkRef,
  serializeDeviceInfo,
  serializeFileMeta,
  serializePrepareUploadRequest,
  serializePrepareUploadResult,
  serializeTransferPlan,
  serializeTransferProgress,
} from './serialization.js';
import {
  arbCapabilities,
  arbChunkRef,
  arbDeviceInfo,
  arbFileMeta,
  arbPrepareUploadRequest,
  arbPrepareUploadResult,
  arbTransferPlan,
  arbTransferProgress,
} from './testing/arbitraries.js';

const RUNS = { numRuns: 200 };

describe('serialization round-trips (Property 6)', () => {
  it('round-trips DeviceInfo', () => {
    fc.assert(
      fc.property(arbDeviceInfo, (info) => {
        expect(parseDeviceInfo(serializeDeviceInfo(info))).toEqual(info);
      }),
      RUNS,
    );
  });

  it('round-trips Capabilities', () => {
    fc.assert(
      fc.property(arbCapabilities, (caps) => {
        expect(parseCapabilities(serializeCapabilities(caps))).toEqual(caps);
      }),
      RUNS,
    );
  });

  it('round-trips TransferPlan', () => {
    fc.assert(
      fc.property(arbTransferPlan, (plan) => {
        expect(parseTransferPlan(serializeTransferPlan(plan))).toEqual(plan);
      }),
      RUNS,
    );
  });

  it('round-trips FileMeta', () => {
    fc.assert(
      fc.property(arbFileMeta, (file) => {
        expect(parseFileMeta(serializeFileMeta(file))).toEqual(file);
      }),
      RUNS,
    );
  });

  it('round-trips ChunkRef (with and without optional sha256)', () => {
    fc.assert(
      fc.property(arbChunkRef, (chunk) => {
        expect(parseChunkRef(serializeChunkRef(chunk))).toEqual(chunk);
      }),
      RUNS,
    );
  });

  it('round-trips PrepareUploadRequest (with and without optional pin)', () => {
    fc.assert(
      fc.property(arbPrepareUploadRequest, (req) => {
        expect(parsePrepareUploadRequest(serializePrepareUploadRequest(req))).toEqual(req);
      }),
      RUNS,
    );
  });

  it('round-trips PrepareUploadResult', () => {
    fc.assert(
      fc.property(arbPrepareUploadResult, (res) => {
        expect(parsePrepareUploadResult(serializePrepareUploadResult(res))).toEqual(res);
      }),
      RUNS,
    );
  });

  it('round-trips TransferProgress', () => {
    fc.assert(
      fc.property(arbTransferProgress, (progress) => {
        expect(parseTransferProgress(serializeTransferProgress(progress))).toEqual(progress);
      }),
      RUNS,
    );
  });
});

describe('canonical serialization is stable and key-sorted', () => {
  it('produces identical JSON regardless of record key insertion order', () => {
    const a = serializePrepareUploadResult({
      sessionId: 's1',
      files: { zeta: 't-z', alpha: 't-a', mu: 't-m' },
      acceptedPlan: {
        mode: 'single-stream',
        chunkSize: 10,
        parallelism: 1,
        chunkProtocolVersion: null,
      },
    });
    const b = serializePrepareUploadResult({
      sessionId: 's1',
      files: { alpha: 't-a', mu: 't-m', zeta: 't-z' },
      acceptedPlan: {
        mode: 'single-stream',
        chunkSize: 10,
        parallelism: 1,
        chunkProtocolVersion: null,
      },
    });
    expect(a).toBe(b);
    expect(a).toContain('"alpha"');
    // Keys must be emitted in sorted order: alpha, mu, zeta.
    expect(a.indexOf('alpha')).toBeLessThan(a.indexOf('mu'));
    expect(a.indexOf('mu')).toBeLessThan(a.indexOf('zeta'));
  });
});

File: apps/android/core/protocol/src/test/kotlin/com/wrapdrive/core/protocol/SerializationTest.kt
package com.wrapdrive.core.protocol

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.kotest.property.Arb
import io.kotest.property.arbitrary.bind
import io.kotest.property.arbitrary.boolean
import io.kotest.property.arbitrary.enum
import io.kotest.property.arbitrary.int
import io.kotest.property.arbitrary.long
import io.kotest.property.arbitrary.orNull
import io.kotest.property.arbitrary.string
import io.kotest.property.checkAll

/**
 * Property 6 — Roundtrip serialization (Kotlin): `parse(serialize(msg)) == msg`
 * for every model, including null optionals.
 *
 * Validates: Requirements 8.1, 7.3, 7.5
 */
class SerializationTest : StringSpec({
    val arbDeviceInfo: Arb<DeviceInfo> =
        Arb.bind(
            Arb.string(1, 32),
            Arb.string(1, 8),
            Arb.string(0, 16).orNull(),
            Arb.enum<DeviceType>().orNull(),
            Arb.string(1, 24),
            Arb.int(1024, 65535),
            Arb.enum<Protocol>(),
            Arb.boolean(),
        ) { alias, version, model, type, fp, port, proto, download ->
            DeviceInfo(alias, version, model, type, fp, port, proto, download)
        }

    val arbFileMeta: Arb<FileMeta> =
        Arb.bind(
            Arb.string(1, 16),
            Arb.string(1, 32),
            Arb.long(0L, Long.MAX_VALUE / 2),
            Arb.string(1, 24),
            Arb.string(64, 64).orNull(),
            Arb.string(0, 16).orNull(),
        ) { id, name, size, type, sha, preview -> FileMeta(id, name, size, type, sha, preview) }

    "DeviceInfo round-trips" {
        checkAll(arbDeviceInfo) { info ->
            ProtocolJson.parseDeviceInfo(ProtocolJson.serialize(info)) shouldBe info
        }
    }

    "FileMeta round-trips" {
        checkAll(arbFileMeta) { file ->
            ProtocolJson.parseFileMeta(ProtocolJson.serialize(file)) shouldBe file
        }
    }

    "TransferPlan round-trips" {
        checkAll(
            Arb.enum<TransferMode>(),
            Arb.long(0L, Long.MAX_VALUE / 2),
            Arb.int(1, 16),
            Arb.string(1, 12).orNull(),
        ) { mode, chunkSize, parallelism, version ->
            val plan = TransferPlan(mode, chunkSize, parallelism, version)
            ProtocolJson.parseTransferPlan(ProtocolJson.serialize(plan)) shouldBe plan
        }
    }
})
